cpp/include/cudf/quantiles.hpp

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+#include <rmm/mr/device_memory_resource.hpp>
+
+namespace cudf {
+
+namespace experimental {
+
+/* @brief Computes a quantile of a given column.
+ *
+ * @param[in] in            Column used to compute quantile.
+ * @param[in] quantile      Requested quantile in range [0, 1].
+ * @param[in] interpolation Interpolation strategy for quantiles lying between
+ *                          two points.
+ * @param[in] mr            Optional. The resource to use for all allocations.
+ * @param[in] stream        Stream on which to perform computations.
+ *
+ * @returns The quantile. Null if column is empty.
+ */
+std::unique_ptr<scalar>
+quantile(column_view const& in,
+         double quantile,
+         interpolation interpolation  = interpolation::LINEAR,
+         bool is_sorted = false,
+         order order = order::ASCENDING,
+         null_order null_order = null_order::AFTER,
+         rmm::mr::device_memory_resource *mr =
+          rmm::mr::get_default_resource(),
+         cudaStream_t stream = 0);
+
+/* @brief Computes a quantile of a given column.
+ *
+ * @param[in] in            Table containing columns used to compute quantiles.
+ * @param[in] quantile      Requested quantile in range [0, 1].
+ * @param[in] interpolation Interpolation strategy for quantiles lying between
+ *                          two points.
+ * @param[in] mr            Optional. The resource to use for all allocations.
+ * @param[in] stream        Stream on which to perform computations.
+ *
+ * @returns Quantiles for each column. Elements are null if columns are empty.
+ */
+std::vector<std::unique_ptr<scalar>>
+quantiles(table_view const& in,
```

Comment:
I’ve Been meaning to ask this for a while. Why should we reduce this surface area? I ask because whenever I try to use these on a column, I’ve had to make tables on the fly and then tediously unpack the result table and std:: move it to a column. Wouldn’t it be better to have that logic in one place and let the column API call the table API internally.

Reply:
I mean to ask in general, not just here

Reply:
> Why should we reduce this surface area?

It's just generally good library design to keep the public API as minimal as possible. Why? Because the more functions you have, the more functions you need to test and maintain. The more time you have to spend on maintenance, the less time there is for developing new features.

In detail, increased surface area means:
- Increase in source of potential bugs
- Need to write more tests
- Increased compile time
- Increased cost of maintenance:
   - Tests, Documentation, Implementation